### DISCUSSION TITLE
Assembly with complex numbers

### DISCUSSION THREAD

**Discussion by gatling-nrl** [kinnala/scikit-fem#1042] [Q&A]
I'm trying to implement a form for $\int uv \ dx$ where u and v are complex.

Naively, I write $\int (u_{re} + iu_{im})(v_{re}+iv_{im}) \ dx$

which leads me to some bilinear form like $i\int u_{re} v_{im} \ dx$

Now I imagine code like 

```python
real_basis = skfem.CellBasis(mesh, element)  #P1 should suffice
imag_basis = skfem.CellBasis(mesh, element)  #P1 should suffice

@skfem.BilinearForm
def form(u, v, _):
    return u * v  # <-- this can't be right....
   # how to deal with those terms that need u from real and v from imag?

form.assemble(....?
```

How do i accomplish this?

EDIT:

It seems like I'm thinking about it all wrong. In the Poisson equation, for example, it seems that the unknown (complex) potential is modelled by real basis functions? I.e.

$u = \sum_n a_n \varphi_n$

the $\varphi_n$ basis functions are real, and only the $a_n$ are complex? In which case I would only need the one real_basis above?

**Comment by kinnala** [comment]
Perhaps this helps: https://github.com/kinnala/scikit-fem/blob/6159aa309cd182578405ee8a84553d9a6d7462de/tests/test_assembly.py#L432

I don't know if it was used anywhere.
